### PR TITLE
MCUX EDMA enhancement

### PR DESCRIPTION
**Pull request by hakehuang**
the former MCUX EDMA driver does not support multi-channel co-exist and minor/major link well, this commit enhances this supporting and adds one test application

### CODE REVIEW

File: drivers/dma/dma_mcux_edma.c [hunk 1]
```diff
@@ -48,8 +48,7 @@ struct dma_mcux_edma_data {
 	struct call_back data_cb[DT_INST_PROP(0, dma_channels)];
 };
 
-#define DEV_CFG(dev)                                                           \
-	((const struct dma_mcux_edma_config *const)dev->config)
+#define DEV_CFG(dev) ((const struct dma_mcux_edma_config *const)dev->config)
```
Comment (MaureenHelm) [inline drivers/dma/dma_mcux_edma.c:51]: Please move the whitespace changes to a separate commit

File: drivers/dma/dma_mcux_edma.c [hunk 2]
```diff
@@ -115,7 +114,8 @@ static void channel_irq(edma_handle_t *handle)
 				tcds_done = 0;
 			}
 		} else {
-			tcds_done = (uint32_t)new_header - (uint32_t)handle->header;
+			tcds_done =
+				(uint32_t)new_header - (uint32_t)handle->header;
```
Comment (MaureenHelm) [inline drivers/dma/dma_mcux_edma.c:118]: This whitespace change should also move. You can put all the whitespace changes together in one commit, just keep them separate from functional changes.

File: tests/drivers/dma/chan_link_transfer/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(chan_blen_transfer)
```
Comment (MaureenHelm) [inline tests/drivers/dma/chan_link_transfer/CMakeLists.txt:5]: should this be `chan_link_transfer`?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: tests/drivers/dma/chan_link_transfer/CMakeLists.txt (hunk 3 above)
Comment (hakehuang) [reply tests/drivers/dma/chan_link_transfer/CMakeLists.txt:5]: yes, just copy out of thinking from chan_blen_transfer, updated

File: drivers/dma/dma_mcux_edma.c (hunk 1 above)
Comment (hakehuang) [reply drivers/dma/dma_mcux_edma.c:51]: revert back

File: drivers/dma/dma_mcux_edma.c (hunk 2 above)
Comment (hakehuang) [reply drivers/dma/dma_mcux_edma.c:118]: revert back

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@mmahadevan108 please review

File: tests/drivers/dma/chan_link_transfer/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+tests:
+  drivers.dma:
+    min_ram: 16
+    depends_on: dma
+    tags: drivers dma
+    platform_allow: frdm_k64f mimxrt1060_evk
```
Comment (nashif) [inline tests/drivers/dma/chan_link_transfer/testcase.yaml:6]: is this really NXP specific? Can we filter on something else here instead to use the same test on other platforms that have the support?

Comment (hakehuang) [reply tests/drivers/dma/chan_link_transfer/testcase.yaml:6]: @nashif , the added feature is from NXP EDMA feature of minor / major loops by using the dest_chaining_en and src_chaining_en, which is related to the RM in below:
image

I did not see any other EDMA drivers defines likely functions. the dest_chaining_en is only defined in the dma.h before. never used. looks like the dma driver is marked as orphaned so shall I raise it to API meeting to get unified?

**Comment by galak** [comment]
@hakehuang can you rebase this on master.

**Comment by hakehuang** [comment]
> @hakehuang can you rebase this on master.
@galak thanks. Rebased. This PR does not contain the dynamic channel allocation, as this requires peripheral driver to support, please check this PR for dynamic dma channel allocation https://github.com/zephyrproject-rtos/zephyr/pull/27737

**Review by mmahadevan108 — APPROVED** [review]
